Loop/Managers/Missed Meal Detection/MealDetectionManager.swift

```diff
@@ -0,0 +1,311 @@
+        self.carbStore = carbStore
+        self.maximumBolus = maximumBolus
+        self.test_currentDate = test_currentDate
+    }
+    
+    // MARK: Meal Detection
+    func hasMissedMeal(insulinCounteractionEffects: [GlucoseEffectVelocity], completion: @escaping (MissedMealStatus) -> Void) {
+        let delta = TimeInterval(minutes: 5)
+
+        let intervalStart = currentDate(timeIntervalSinceNow: -MissedMealSettings.maxRecency)
+        let intervalEnd = currentDate(timeIntervalSinceNow: -MissedMealSettings.minRecency)
+        let now = self.currentDate
+
+        carbStore.getGlucoseEffects(start: intervalStart, end: now, effectVelocities: insulinCounteractionEffects) {[weak self] result in
+            guard
+                let self = self,
+                case .success((let carbEntries, let carbEffects)) = result
+            else {
+                if case .failure(let error) = result {
+                    self?.log.error("Failed to fetch glucose effects to check for missed meal: %{public}@", String(describing: error))
+                }
+
+                completion(.noMissedMeal)
+                return
+            }
+            
+            /// Compute how much of the ICE effect we can't explain via our entered carbs
+            /// Effect caching inspired by `LoopMath.predictGlucose`
+            var effectValueCache: [Date: Double] = [:]
+            let unit = HKUnit.milligramsPerDeciliter
+
+            /// Carb effects are cumulative, so we have to subtract the previous effect value
+            var previousEffectValue: Double = carbEffects.first?.quantity.doubleValue(for: unit) ?? 0
+
+            /// Counteraction effects only take insulin into account, so we need to account for the carb effects when computing the unexpected deviations
+            for effect in carbEffects {
+                let value = effect.quantity.doubleValue(for: unit)
+                /// We do `-1 * (value - previousEffectValue)` because this will compute the carb _counteraction_ effect
+                effectValueCache[effect.startDate] = (effectValueCache[effect.startDate] ?? 0) +  -1 * (value - previousEffectValue)
+                previousEffectValue = value
+            }
+
+            let processedICE = insulinCounteractionEffects
+                .filterDateRange(intervalStart, now)
+                .compactMap {
+                    /// Clamp starts & ends to `intervalStart...now` since our algorithm assumes all effects occur within that interval
+                    let start = max($0.startDate, intervalStart)
+                    let end = min($0.endDate, now)
+
+                    guard let effect = $0.effect(from: start, to: end) else {
+                        let item: GlucoseEffect? = nil // FIXME: we get a compiler error if we try to return `nil` directly
+                        return item
+                    }
+
+                    return GlucoseEffect(startDate: effect.startDate.dateFlooredToTimeInterval(delta),
```

Comment:
It might not make a big impact in meal detection (might make meals appear to be detected earlier in time than the effects that are triggering the detection). Just thought I'd point out that it's not strictly correct. Dealing with this properly might be complex, since you're wanting to ultimately subtract entered carb effects from ICE, which are both on different timelines; carb effects are on a simulation timeline, and ICE is on the reported glucose timeline.

Reply:
I agree that it's not ideal to have to translate the ICE into the type of effect that carb effects can be subtracted from. I just played with using a ceiled endDate to assign the GlucoseEffect `startDate` (so for your example, the startDate would be 1:55pm) and ran the unit tests. It actually seems that this change leads to an earlier meal start time being assigned. 

From a safety perspective, which behavior would be preferred - potentially having the missed meal time be too early or too late? I would think too early might be preferred (1) carb digestion & CGM lag already means we're lagging behind and (2) because there would be less remaining effect for Loop to dose for

Reply:
Yes, I think making sure we cover more of the time that carbs could have been absorbing is the more conservative approach. This feature in general will turn off RC unless the user enters more carbs than detected, so I imagine it mostly acting as a nudge towards less delivery, even if the start time isn't detected precisely.

Again, I don't think this affects things too much.